rusk/src/lib/chain/rusk.rs

```diff
@@ -325,22 +354,52 @@ impl Rusk {
         Ok(session)
     }
 
-    pub(crate) fn set_current_commit(&self, commit: [u8; 32]) {
+    pub(crate) fn set_current_commit(
+        &self,
+        commit: [u8; 32],
+        block_height: u64,
+    ) -> Result<()> {
         let mut tip = self.tip.write();
+
+        // If the current block height is a multiple of the epoch, we need to
+        // write the commit to the epoch commit file, and set the commit in the
+        // tip.
+        if block_height % EPOCH == 0 {
+            let epoch_commit_path = to_rusk_epoch_id_path(&self.dir);
+            write_commit_id(commit, epoch_commit_path)?;
+            tip.epoch = Some(commit);
+        }
+
         tip.current = commit;
+
+        Ok(())
     }
 
-    pub(crate) fn set_base_and_delete(&self, commit: [u8; 32]) {
+    pub(crate) fn set_base_and_delete(
+        &self,
+        commit: [u8; 32],
+        block_height: u64,
+    ) -> Result<()> {
         let mut tip = self.tip.write();
 
+        // If the current block height is a multiple of the epoch, we need to
+        // write the commit to the epoch commit file, and set the commit in the
+        // tip.
+        if block_height % EPOCH == 0 {
+            let epoch_commit_path = to_rusk_epoch_id_path(&self.dir);
+            write_commit_id(commit, epoch_commit_path)?;
+            tip.epoch = Some(commit);
+        }
+
         let current_commit = tip.current;
         let base_commit = tip.base;
 
         tip.current = commit;
         tip.base = commit;
 
         // We will delete all commits except the previous base commit, the
-        // previous current commit and the new commit.
+        // previous current commit, the new commit and the epoch commit - if it
+        // exists.
```

Comment:
This seems like really strange behavior. Why is this even the case? I would expect that provisioners would simply be queried once, and have nothing to do with the previous block.

Reply:
It's a caveaut of the consensus.
For example, when your node is at block with height = H, and you're accepting the next block with height=H+1, you have to verify the votes against provisioners of the current state (height=H). Additionally, since that the block with height = H+1 carries the winning certificate of the block with height=H you have to verify it against provisioner of the state at height=H-1 (the ones that votes for block with height=H)

That's why we carries ContextProvisioners that have both current and prev_provisioners.

When a node needs to start (or when falling back) it needs to query for both provisioners set.

This means that when we sync from somewhere, we need to have both states.
That's why I was suggesting to keep track of prev_provisioners directly in the stake contract, so to have only one state to derive both info

@goshawk-3 am I missing something?

Reply:
Correct. As per [block header definition](https://github.com/dusk-network/dusk-protocol/blob/main/consensus/chain-management/README.md), we always verify `PrevBlockCertificate` on accepting a block. 

On reverting, we try to retrieve two provisioners lists.
See also https://github.com/dusk-network/rusk/blob/c36956bade1576b770f35466bfe778043402b499/node/src/chain/acceptor.rs#L368

I accept that RevertToEpoch will fail on the above line.

Reply:
I suppose this means we always need the state of two consecutive blocks for any reversion target? This is not even true right now with reverting to finalized blocks BTW.

Reply:
That's a valid comment. Maybe the commit associated to `prev_header.state_hash` is not deleted at all.

Reply:
> I suppose this means we always need the state of two consecutive blocks for any reversion target? This is not even true right now with reverting to finalized blocks BTW.

When we revert to last finalized block, we ask indeed for previous block.
The trick is inside the "finalize" that doesn't delete the previous state root